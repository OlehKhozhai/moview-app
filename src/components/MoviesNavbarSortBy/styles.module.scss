@import 'assets/styles/mixins/colors';

.root {
  display: flex;
  align-items: center;
}

.label {
  color: $primary-dark-color;
  text-transform: uppercase;
  margin-right: 40px;
}
