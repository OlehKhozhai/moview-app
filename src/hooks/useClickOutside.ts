import { RefObject, useEffect, useState } from 'react';

type Props = {
  ref: RefObject<HTMLElement>;
  handler?: keyof GlobalEventHandlersEventMap;
  closeOnOutsideClick?: boolean;
  callback?: () => void;
};

const useClickOutside = ({
  ref,
  callback,
  handler = 'click',
  closeOnOutsideClick = true,
}: Props) => {
  const [isClickOutside, setClickOutside] = useState(false);

  useEffect(() => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const handleClick = (e: any) => {
      const targetIsValid = e.target && document.contains(e.target);
      if (ref && ref.current && targetIsValid) {
        const isClickOnRef = ref.current.contains(e.target);

        if (!isClickOnRef && closeOnOutsideClick) {
          setClickOutside(true);
          if (callback) {
            callback();
          }
        }
      }
    };

    if (ref?.current) {
      document.addEventListener(handler, handleClick);
    }
    return () => {
      document.removeEventListener(handler, handleClick);
    };
  }, [handler, ref, closeOnOutsideClick, callback]);

  return [isClickOutside];
};

export default useClickOutside;
