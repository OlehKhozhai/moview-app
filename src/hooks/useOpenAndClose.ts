import React from 'react';

type Props = { onCloseCallback?: () => void } | undefined;

const useOpenAndClose = (props: Props = {}) => {
  const { onCloseCallback } = props;
  const [isOpen, setIsOpen] = React.useState(false);

  const handleClose = React.useCallback(() => {
    if (onCloseCallback) {
      onCloseCallback();
    }

    setIsOpen(false);
  }, [onCloseCallback]);

  const handleOpen = React.useCallback(() => {
    setIsOpen(true);
  }, []);

  const handleToggle = React.useCallback(
    (e?: React.MouseEvent<HTMLElement>) => {
      if (e) {
        e.stopPropagation();
      }

      setIsOpen((prevState) => {
        if (!prevState && onCloseCallback) {
          onCloseCallback();
        }

        return !prevState;
      });
    },
    [onCloseCallback]
  );

  return {
    isOpen,
    onClose: handleClose,
    onOpen: handleOpen,
    onToggle: handleToggle,
  };
};

export default useOpenAndClose;
