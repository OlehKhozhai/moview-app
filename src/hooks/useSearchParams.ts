import { useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';

import { DROPDOWN_OPTIONS } from 'config';

const useSearchParams = (): string => {
  const [params, setParams] = useState('');

  const {
    replace: historyReplace,
    location: { search: searchQuery },
  } = useHistory();

  useEffect(() => {
    const searchParams = new URLSearchParams(searchQuery);
    const activeTab = searchParams.get('search');
    const sortBy = searchParams.get('sortBy');

    if (!activeTab) {
      searchParams.set('searchBy', 'genres');
      searchParams.set('search', '');
    }

    if (!sortBy) {
      searchParams.set('sortBy', DROPDOWN_OPTIONS[0]);
    }

    historyReplace({ search: searchParams.toString() });
    setParams(searchParams.toString());
  }, [searchQuery, historyReplace]);

  return params;
};

export default useSearchParams;
