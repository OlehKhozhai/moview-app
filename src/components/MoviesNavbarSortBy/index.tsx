import React from 'react';
import cn from 'classnames';
import { useHistory } from 'react-router-dom';

import Dropdown from 'components/_common/Dropdown';
import { DROPDOWN_OPTIONS } from 'config';
import useOpenAndClose from 'hooks/useOpenAndClose';
import styles from './styles.module.scss';

type MoviesNavbarSortByProps = {
  searchParams: string;
  className?: string;
};

const MoviesNavbarSortBy: React.FC<MoviesNavbarSortByProps> = ({ searchParams, className }) => {
  const { replace: historyReplace } = useHistory();
  const { isOpen, onToggle, onClose } = useOpenAndClose();

  const newSearchParams = React.useMemo(() => new URLSearchParams(searchParams), [searchParams]);
  const sortBy = newSearchParams.get('sortBy') || '';

  const handleOptionClick = React.useCallback(
    (option: string) => {
      newSearchParams.set('sortBy', option);
      historyReplace({ search: newSearchParams.toString() });
      onClose();
    },
    [newSearchParams, onClose, historyReplace]
  );

  return (
    <div className={cn(styles.root, className)}>
      <span className={styles.label}>Sort By</span>

      <Dropdown
        options={DROPDOWN_OPTIONS}
        value={sortBy}
        isOpen={isOpen}
        onToggle={onToggle}
        onClose={onClose}
        onOptionClick={handleOptionClick}
      />
    </div>
  );
};

export default React.memo(MoviesNavbarSortBy);
