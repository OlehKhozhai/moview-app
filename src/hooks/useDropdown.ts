import React from 'react';

import useOpenAndClose from 'hooks/useOpenAndClose';

type Props =
  | {
      defaultValue?: string;
    }
  | undefined;

const useDropdown = (props: Props = {}) => {
  const { defaultValue } = props;
  const [value, setValue] = React.useState(defaultValue || '');
  const { isOpen, onToggle, onClose } = useOpenAndClose();

  const handleOptionClick = React.useCallback(
    (option: string) => {
      setValue(option);
      onClose();
    },
    [onClose]
  );

  return {
    value,
    onOptionClick: handleOptionClick,
    isDropdownVisible: isOpen,
    onToggleDropdownVisibility: onToggle,
    onCloseDropdown: onClose,
  };
};

export default useDropdown;
