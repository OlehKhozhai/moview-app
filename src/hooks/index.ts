import useOpenAndClose from './useOpenAndClose';
import useDropdown from './useDropdown';
import useClickOutside from './useClickOutside';

export default { useOpenAndClose, useDropdown, useClickOutside };
