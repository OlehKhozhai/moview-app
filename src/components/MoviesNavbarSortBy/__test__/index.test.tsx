import React from 'react';
import { create } from 'react-test-renderer';

import hooks from 'hooks';
import MoviesNavbarSortBy from '../index';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn().mockReturnValue({ replace: jest.fn() }),
}));

describe('MoviesNavbarSortBy component', () => {
  test('should render MoviesNavbarSortBy component by default', () => {
    const component = create(<MoviesNavbarSortBy searchParams="" />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should render MoviesNavbarSortBy component with open dropdown', () => {
    jest.spyOn(hooks, 'useOpenAndClose').mockReturnValueOnce({
      isOpen: true,
      onOpen: jest.fn(),
      onClose: jest.fn(),
      onToggle: jest.fn(),
    });

    const component = create(<MoviesNavbarSortBy searchParams="" />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
