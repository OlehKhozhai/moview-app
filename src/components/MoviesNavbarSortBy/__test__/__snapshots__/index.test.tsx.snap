// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesNavbarSortBy component should render MoviesNavbarSortBy component by default 1`] = `
<div
  className="root"
>
  <span
    className="label"
  >
    Sort By
  </span>
  <div
    className="root"
  >
    <span
      className="threeDots"
      onClick={[Function]}
    >
      ...
    </span>
  </div>
</div>
`;

exports[`MoviesNavbarSortBy component should render MoviesNavbarSortBy component with open dropdown 1`] = `
<div
  className="root"
>
  <span
    className="label"
  >
    Sort By
  </span>
  <div
    className="root"
  >
    <span
      className="threeDots"
      onClick={[MockFunction]}
    >
      ...
    </span>
    <ul
      className="dropdown"
    >
      <li
        className="root"
        onClick={[Function]}
      >
        release date
      </li>
      <li
        className="root"
        onClick={[Function]}
      >
        rating
      </li>
    </ul>
  </div>
</div>
`;
